<a href="https://colab.research.google.com/github/Amjad-199/Python_Practicals/blob/main/PythonIntro_part2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Python and Google Colab (Part 2)

In Part 1 your code ran straight down the page, and every line ran exactly once.

This session gives your code three new powers: it can **make a decision**, it can **repeat itself**, and it can be **packaged up under a name and reused**. We finish with NumPy, which does arithmetic on a whole column of numbers at once.


Part 2 learning objectives:

1. To make your code choose between alternatives with if, elif and else
2. To repeat work automatically with for loops and while loops
3. To write your own functions, with arguments and return values
4. To meet lambda functions and the map / filter / reduce family
5. To use NumPy arrays to work on a whole set of numbers at once

This notebook builds on Part 1, but starts from scratch, so nothing from last week needs to be loaded. Exercises are marked with a pencil.


---



## 1. Making decisions: if, elif and else


### 1.1 The if statement

So far every line of your code has run. An if statement runs a block of code only when a condition is true, using the True/False comparisons you met in Part 1.

In [ ]:
temperature = 25

if temperature > 20:
    print("It is warm today")

It is warm today


Change the 25 to 15 and run it again. Nothing is printed, because the condition is False and the indented block is skipped.

The shape of an if statement is always the same:

```python
if some_condition_that_is_true_or_false:
    do this
    and this
and this happens either way, because it is not indented
```

### 1.2 Indentation is not decoration

Two things mark the block of code that belongs to an if:

1. A colon `:` at the end of the if line. Forget it and you get a `SyntaxError`.
2. Indentation: four spaces at the start of every line in the block. The block ends when the indentation goes back out again.

Other languages use curly brackets for this. Python uses the layout itself, which means well indented code and correct code are the same thing. Colab indents for you after a colon; the **Tab** key indents, and **Shift + Tab** removes indentation.



### 1.3 else and elif

An `else` block says what to do when the condition was False:

In [ ]:
a = 1

if a > 5:
    print("This should not happen.")
else:
    print("This should happen.")

This should happen.


And `elif`, short for “else if”, lets you test a second or third condition. Python works down the chain, runs the block belonging to the **first** condition that is true, then skips the rest:

In [ ]:
rating = 3

if rating >= 4:
    print("Positive feedback")
elif rating == 3:
    print("Neutral feedback")
else:
    print("Negative feedback")

Neutral feedback


A longer chain turns a number into a category, which you will do a lot with real data:

In [ ]:
score = 72

if score >= 70:
    grade = "Distinction"
elif score >= 60:
    grade = "Merit"
elif score >= 50:
    grade = "Pass"
else:
    grade = "Fail"

print(f"A score of {score} is a {grade}")

A score of 72 is a Distinction


Order matters. Because Python stops at the first true condition, putting `score >= 50` first would grade everybody who passed as a Pass. You may have as many `elif` blocks as you like, and at most one `else`, which must come last.

### 1.4 Conditions on collections

The `in` keyword is often what you want inside an if:

In [ ]:
countries = ["USA", "GBR", "CHN"]

if "GBR" in countries:
    print("Great Britain is in the table")

Great Britain is in the table


Conditions combine with and / or / not just as before:

In [ ]:
stock = {"apples": True, "pears": False}
item = "pears"

if item in stock and stock[item]:
    print(f"{item} are in stock")
else:
    print(f"sorry, no {item} today")

sorry, no pears today


> **&#9998; Exercise.** Write a program that takes a variable holding the time of day ("morning", "afternoon" or "evening") and prints a suitable greeting for each. Make sure it also does something sensible for a value you did not expect.

In [ ]:
# your answer here

In [ ]:
time_of_day = "morning"  # You can change this value to test different times

if time_of_day == "morning":
    print("Good morning!")
elif time_of_day == "afternoon":
    print("Good afternoon!")
elif time_of_day == "evening":
    print("Good evening!")
else:
    print("Hello! It's a lovely day.") # Handles unexpected values

Good morning!


> **&#9998; Exercise.** A customer rates a service from 1 to 5. Write an if/elif/else chain that prints "positive" for 4 or 5, "neutral" for 3, "negative" for 1 or 2, and "invalid rating" for anything else. Test it with several values.

In [ ]:
# your answer here

In [ ]:
def get_rating_sentiment(rating):
    if rating == 4 or rating == 5:
        print(f"Rating {rating}: Positive")
    elif rating == 3:
        print(f"Rating {rating}: Neutral")
    elif rating == 1 or rating == 2:
        print(f"Rating {rating}: Negative")
    else:
        print(f"Rating {rating}: Invalid rating")

# Test cases
print("--- Testing ratings ---")
get_rating_sentiment(5)
get_rating_sentiment(4)
get_rating_sentiment(3)
get_rating_sentiment(2)
get_rating_sentiment(1)
get_rating_sentiment(0)  # Invalid
get_rating_sentiment(6)  # Invalid
get_rating_sentiment("three") # Invalid (non-numeric)

--- Testing ratings ---
Rating 5: Positive
Rating 4: Positive
Rating 3: Neutral
Rating 2: Negative
Rating 1: Negative
Rating 0: Invalid rating
Rating 6: Invalid rating
Rating three: Invalid rating


> **&#9998; Exercise.** Write a check that prints a person's access level from their role: "admin" has full access, "user" has limited access, "guest" has none.

In [ ]:
# your answer here

In [ ]:
role = "admin"  # You can change this value to test different roles

if role == "admin":
    print(f"Role: {role}, Access Level: Full access")
elif role == "user":
    print(f"Role: {role}, Access Level: Limited access")
elif role == "guest":
    print(f"Role: {role}, Access Level: No access")
else:
    print(f"Role: {role}, Access Level: Unknown role, no access granted.")


Role: admin, Access Level: Full access


## 2. Repeating work: the for loop



### 2.1 Doing something to every item

To print a message for each of twenty countries you could write twenty print statements, or one loop. A for loop takes a collection and runs the same block once for each item in it.

In [ ]:
fruits = ["apple", "banana", "orange"]

for fruit in fruits:
    print("I love", fruit)

I love apple
I love banana
I love orange


Read it as English: “for each fruit in fruits, print it”. The name `fruit` is one you choose; it holds a different item each time round the loop. Note the punctuation you now recognise: a colon at the end of the line, and an indented block underneath.

A loop works over anything with items in it, including a string, where the items are the individual characters:

In [ ]:
for letter in "data":
    print(letter.upper())

D
A
T
A


### 2.2 Counting with range()

To repeat something a fixed number of times, `range()` produces a sequence of whole numbers for the loop to walk through:

In [ ]:
for i in range(5):
    print(i)

0
1
2
3
4


Two things to notice. It starts at 0, and it stops **before** the number you gave, the same “up to but not including” rule as slicing in Part 1. Given two numbers, range starts at the first and stops before the second; given three, the third is the step:

In [ ]:
for i in range(1, 6):
    print(i, i ** 2)

1 1
2 4
3 9
4 16
5 25


In [ ]:
for i in range(0, 10, 2):
    print(i)

0
2
4
6
8


### 2.3 Building up an answer

A common pattern is to set up a variable before the loop and change it a little on each pass. Here is a running total:

In [ ]:
total = 0

for value in [1.0, 2.3, 0.2, 1.1]:
    total = total + value

print("total is", total)

total is 4.6


The same pattern builds a new list, collecting a result for each item:

In [ ]:
names = ["Tom", "Deep", "Harry"]
lengths = []

for name in names:
    lengths.append(len(name))

print(lengths)

[3, 4, 5]


Python has a shorthand for that pattern, called a **list comprehension**. It puts the loop inside the square brackets of the new list, and reads almost like the English description: “the length of each name, for each name in names”:

In [ ]:
lengths = [len(name) for name in names]
print(lengths)

[3, 4, 5]


You can add a condition on the end to keep only some of the items:

In [ ]:
print([n for n in range(1, 11) if n % 2 == 0])

[2, 4, 6, 8, 10]


Write the loop version first while the pattern is new. The comprehension is a shorthand for it, not a different idea, but you will see comprehensions everywhere in other people's code, so it is worth being able to read one.

### 2.4 Looping over a dictionary

Looping over a dictionary gives you its keys:

In [ ]:
medals = {"USA": 121, "GBR": 67, "CHN": 70}

for country in medals:
    print(country)

USA
GBR
CHN


Usually you want the key and the value together, and `.items()` hands you both at once. Notice the two variable names before the `in`:

In [ ]:
for country, count in medals.items():
    print(f"{country} won {count} medals")

USA won 121 medals
GBR won 67 medals
CHN won 70 medals


A related trick for lists: `enumerate()` gives you the position alongside the item, which saves you counting by hand:

In [ ]:
for position, fruit in enumerate(fruits):
    print(position, fruit)

0 apple
1 banana
2 orange


### 2.5 Loops inside loops

A loop can contain another loop. The inner loop runs all the way through on every pass of the outer one, which is what you need to walk over a table of rows and columns:

In [ ]:
grid = [[1, 2, 3], [4, 5, 6]]

for row in grid:
    for value in row:
        print(value, end=" ")
    print()

1 2 3 
4 5 6 


Watch the indentation here, because it is doing all the work. The `print(value...)` line is indented twice, so it belongs to the inner loop. The bare `print()` is indented once, so it belongs to the outer one and gives a line break at the end of each row. (`end=" "` tells print to finish with a space instead of a new line.)

### 2.6 Stopping early: break and continue

Two keywords change the flow of a loop. `break` leaves the loop altogether; `continue` skips the rest of this pass and goes straight on to the next one:

In [ ]:
for n in range(1, 11):
    if n == 5:
        continue     # skip 5
    if n > 7:
        break        # stop completely once past 7
    print(n)

1
2
3
4
6
7


> **&#9998; Exercise.** Use a for loop and range() to print the numbers 1 to 20, but only the even ones. (Hint: the % operator from Part 1 tells you the remainder.)

In [ ]:
# your answer here

In [ ]:
for number in range(1, 21):
    if number % 2 == 0:
        print(number)

2
4
6
8
10
12
14
16
18
20


> **&#9998; Exercise.** Given the list [4, 17, 3, 28, 9, 41, 6], use a loop to find the largest value without using max(). Then check your answer against max().

In [ ]:
# your answer here

In [ ]:
numbers = [4, 17, 3, 28, 9, 41, 6]

# Find the largest value using a loop
largest_value_loop = numbers[0] # Initialize with the first element

for number in numbers:
    if number > largest_value_loop:
        largest_value_loop = number

print(f"Largest value found with loop: {largest_value_loop}")

# Check the answer against max()
largest_value_max = max(numbers)
print(f"Largest value found with max(): {largest_value_max}")

# Verify if they are the same
if largest_value_loop == largest_value_max:
    print("The values match!")
else:
    print("There is a mismatch.")

Largest value found with loop: 41
Largest value found with max(): 41
The values match!


> **&#9998; Exercise.** Take the dictionary `{"USA": [46, 37, 38], "GBR": [27, 23, 17], "CHN": [26, 18, 26]}` and loop over it to print one line per country in the form “USA won 121 medals in total”.

In [ ]:
medals_data = {"USA": [46, 37, 38], "GBR": [27, 23, 17], "CHN": [26, 18, 26]}

for country, medal_counts in medals_data.items():
    total_medals = sum(medal_counts)
    print(f"{country} won {total_medals} medals in total")

USA won 121 medals in total
GBR won 67 medals in total
CHN won 70 medals in total


> **&#9998; Exercise.** Here is a list of lists of sentences:

In [ ]:
text_data = [
    ["The quick brown fox jumps over the lazy dog",
     "A stitch in time saves nine"],
    ["All that glitters is not gold", "Fortune favours the bold"],
    ["An apple a day keeps the doctor away",
     "Necessity is the mother of invention"],
]

# 1. print every word on its own line   (hint: .split())
# 2. count how many times the word "is" appears in total
# 3. find the longest word in each inner list

In [ ]:
text_data = [
    ["The quick brown fox jumps over the lazy dog",
     "A stitch in time saves nine"],
    ["All that glitters is not gold", "Fortune favours the bold"],
    ["An apple a day keeps the doctor away",
     "Necessity is the mother of invention"],
]

# 1. Print every word on its own line
print("--- Every word on its own line ---")
for sentence_list in text_data:
    for sentence in sentence_list:
        for word in sentence.split():
            print(word)

# 2. Count how many times the word "is" appears in total
print("\n--- Counting 'is' ---")
is_count = 0
for sentence_list in text_data:
    for sentence in sentence_list:
        for word in sentence.split():
            if word.lower() == "is":
                is_count += 1
print(f"The word 'is' appears {is_count} times.")

# 3. Find the longest word in each inner list
print("\n--- Longest word in each inner list ---")
for i, sentence_list in enumerate(text_data):
    longest_word_in_sublist = ""
    for sentence in sentence_list:
        for word in sentence.replace('.', '').replace(',', '').split(): # Clean punctuation for accurate length
            if len(word) > len(longest_word_in_sublist):
                longest_word_in_sublist = word
    print(f"Longest word in inner list {i+1}: '{longest_word_in_sublist}' (length: {len(longest_word_in_sublist)})")

--- Every word on its own line ---
The
quick
brown
fox
jumps
over
the
lazy
dog
A
stitch
in
time
saves
nine
All
that
glitters
is
not
gold
Fortune
favours
the
bold
An
apple
a
day
keeps
the
doctor
away
Necessity
is
the
mother
of
invention

--- Counting 'is' ---
The word 'is' appears 2 times.

--- Longest word in each inner list ---
Longest word in inner list 1: 'stitch' (length: 6)
Longest word in inner list 2: 'glitters' (length: 8)
Longest word in inner list 3: 'Necessity' (length: 9)


## 3. The while loop

A for loop runs once for each item in a collection, so you know before it starts how many passes there will be. A while loop keeps going for as long as a condition stays true, which is what you want when you cannot say in advance how many repeats you need.

In [ ]:
a = 0

while a < 5:
    a = a + 1
    print(a)

1
2
3
4
5


Read it as: “as long as a is less than 5, add one to a and print it”. Python checks the condition, runs the block, checks again, and stops the moment the condition is False.

In [ ]:
countdown = 3

while countdown > 0:
    print(countdown)
    countdown = countdown - 1

print("Lift off!")

3
2
1
Lift off!


> **Note:** something inside the loop must eventually make the condition False, or the loop runs forever. If that happens, the cell keeps going with a spinning icon; click the stop button (the square where the run button was) to interrupt it. It happens to everyone at least once.

For most data work you will reach for a for loop far more often, because you almost always have a collection to walk through. Use a while loop when the stopping point depends on something you compute as you go.

## 4. Writing your own functions

You have been using functions since your first line of Python: `print()`, `len()`, `round()`, `sum()`. Each one packages a job under a name so you can ask for it without knowing how it works. Now you will write your own.

The reason is a principle you will hear often: don't repeat yourself. Whenever you have copied and pasted a few lines and changed one value, those lines want to be a function. Then a fix happens in one place instead of five.



### 4.1 Defining and calling

A function definition starts with the keyword `def`, then the name you are giving it, then brackets, then a colon. The body of the function is the indented block underneath:

In [ ]:
def greet():
    print("Hello")
    print("Welcome to the practical")

Running that cell prints nothing. It only teaches Python what the word “greet” means. To run it, you call it: the name followed by brackets.

In [ ]:
greet()

Hello
Welcome to the practical


> **Note:** a function must be defined before you call it. If you call it first, Python has not learned the name yet and you get a `NameError`. Keep your definitions near the top of the notebook, or in the cell just above where you first use them.

### 4.2 Giving a function something to work with

A function is far more useful if you can hand it a value to work on. The names in the brackets of the definition are called parameters; the values you hand over when you call it are called arguments.

In [ ]:
def greet_person(name):
    print(f"Hello {name}, welcome to the practical")

greet_person("Adriana")
greet_person("Billy")

Hello Adriana, welcome to the practical
Hello Billy, welcome to the practical


One definition, used as many times as you like with different values. You can pass a variable rather than a literal value, and the names need not match:

In [ ]:
student = "Caroline"
greet_person(student)

Hello Caroline, welcome to the practical


Inside the function that value is known as `name`, whatever it was called outside. What matters is the **position** of each argument, not its name, as this deliberately broken example shows:

In [ ]:
def describe_person(first_name, last_name, age):
    print(f"{first_name} {last_name} is {age} years old")

describe_person("Ada", "Lovelace", 36)

Ada Lovelace is 36 years old


In [ ]:
describe_person(36, "Ada", "Lovelace")     # arguments in the wrong order

36 Ada is Lovelace years old


No error, just nonsense, which is worse. Getting the order right is your job.

### 4.3 Getting an answer back: return

So far our functions have printed something. Much more often you want a function to hand back a **value** that the rest of your code can use. That is what `return` does:

In [ ]:
def total_medals(gold, silver, bronze):
    total = gold + silver + bronze
    return total

usa = total_medals(46, 37, 38)
print("USA won", usa, "medals")

USA won 121 medals


Because it returns a value, the call can be used anywhere a value can be used: inside a print, in a calculation, or stored in a variable.

In [ ]:
print(total_medals(27, 23, 17))

67


The difference between printing and returning matters. A function that prints puts text on the screen and hands back nothing (in fact it hands back `None`). A function that returns hands back something you can go on using. If you cannot use a function's answer, check that you wrote `return` and not `print`.

A function stops the moment it hits a `return`, which is useful for dealing with an awkward case first:

In [ ]:
def average(values):
    if len(values) == 0:
        return 0
    return sum(values) / len(values)

print(average([1.0, 2.3, 0.2, 1.1]))
print(average([]))

1.15
0


Here is a longer one using everything from this session: a loop, an if, and a return.

In [ ]:
def count_vowels(text):
    count = 0
    for letter in text.lower():
        if letter in "aeiou":
            count = count + 1
    return count

print(count_vowels("Introduction to Data Science"))

11


### 4.4 Default values

A parameter can be given a default, which makes it optional when the function is called:

In [ ]:
def greet_with(name, greeting="Hello"):
    print(f"{greeting} {name}")

greet_with("Ada")
greet_with("Ada", "Good morning")
greet_with("Ada", greeting="Welcome")

Hello Ada
Good morning Ada
Welcome Ada


The last line names the argument explicitly. Naming arguments makes a call easier to read when there are several, and the order no longer matters. You saw this style in `sorted(scores, reverse=True)` in Part 1, and it is everywhere in the data libraries we meet in Part 3.

### 4.5 Where variables live: local and global

A variable created inside a function exists only inside that function. It is created when the function runs and forgotten when it finishes. That is a feature, not a limitation: it means a function cannot accidentally interfere with the rest of your code.

In [ ]:
def shout(text):
    loud = text.upper()
    return loud

print(shout("hello"))
#print(loud)     # deliberate error

HELLO


A variable created outside any function is global, and a function can read it:

In [ ]:
message = "I am available everywhere"

def show_message():
    print(message)

show_message()

I am available everywhere


Keep this to a minimum. A function that takes everything it needs as arguments and hands its answer back with return is one you can read, test and reuse on its own. A function that quietly depends on variables defined elsewhere is much harder to trust.

> **&#9998; Exercise.** Write a function that takes a list of numbers and returns the sum of the positive ones only. Test it on [1, 2, 3, -1, 4, 5, -4].

In [ ]:
# your answer here

In [ ]:
def sum_positive_numbers(numbers_list):
    total_positive = 0
    for number in numbers_list:
        if number > 0:
            total_positive += number
    return total_positive

# Test the function
test_list = [1, 2, 3, -1, 4, 5, -4]
result = sum_positive_numbers(test_list)
print(f"The sum of positive numbers in {test_list} is: {result}")

The sum of positive numbers in [1, 2, 3, -1, 4, 5, -4] is: 15


> **&#9998; Exercise.** Write a function that takes a string and returns the number of words in it. Then write another that returns the longest word. Test both on a sentence of your choosing.

In [ ]:
# your answer here

In [ ]:
def count_words(text):
    words = text.split()
    return len(words)

def find_longest_word(text):
    # Remove common punctuation to ensure accurate word length
    cleaned_text = text.replace('.', '').replace(',', '').replace('!', '').replace('?', '')
    words = cleaned_text.split()

    if not words:
        return ""

    longest_word = ""
    for word in words:
        if len(word) > len(longest_word):
            longest_word = word
    return longest_word

# Test both functions
sentence = "Python is a powerful programming language!"

num_words = count_words(sentence)
print(f"The sentence: '{sentence}'")
print(f"Number of words: {num_words}")

longest = find_longest_word(sentence)
print(f"Longest word: '{longest}' (length: {len(longest)})")

The sentence: 'Python is a powerful programming language!'
Number of words: 6
Longest word: 'programming' (length: 11)


> **&#9998; Exercise.** Write a function that takes a character and a string and returns how many times that character appears in the string.

In [ ]:
# your answer here

In [ ]:
def count_char(char, text):
    count = 0
    for c in text:
        if c == char:
            count += 1
    return count

print(count_char("a", "banana"))  # 3
print(count_char("z", "banana"))  # 0

3
0


> **&#9998; Exercise.** Write a function `medal_summary(country, gold, silver, bronze)` that returns a sentence such as “GBR won 67 medals in total, including 27 gold”. Then loop over a dictionary of three countries and print a line for each.

In [ ]:
# your answer here
def medal_summary(country, gold, silver, bronze):
    """Return a one-sentence summary of a country's medal haul."""
    total = gold + silver + bronze
    return f"{country} won {total} medals in total, including {gold} gold"

# Rio 2016 medal counts: (gold, silver, bronze)
medals = {
    "USA": (46, 37, 38),
    "GBR": (27, 23, 17),
    "CHN": (26, 18, 26),
}

for country, (gold, silver, bronze) in medals.items():
    print(medal_summary(country, gold, silver, bronze))

USA won 121 medals in total, including 46 gold
GBR won 67 medals in total, including 27 gold
CHN won 70 medals in total, including 26 gold


## 5. Lambda: a function in one line

Sometimes a function is so small that giving it a name and three lines of definition feels like too much ceremony. A **lambda** is a one-line function with no name, for exactly those cases.

The shape is `lambda arguments: expression`. There is no `return`: the value of the single expression is handed back automatically.

In [ ]:
square = lambda x: x * x
print(square(8))

64


That is the same as, and no better than, writing it out in full:

In [ ]:
def square(x):
    return x * x

So why bother? Because a lambda can be written *where it is needed*, as an argument to another function. The classic case is telling `sorted()` what to sort by. Its `key` argument takes a function, which is applied to each item to work out its sort position:

In [ ]:
words = ["banana", "fig", "cherry", "date"]

print(sorted(words))                      # alphabetical, the default
print(sorted(words, key=lambda w: len(w)))   # by length instead

['banana', 'cherry', 'date', 'fig']
['fig', 'date', 'banana', 'cherry']


In [ ]:
people = [("Ada", 36), ("Alan", 41), ("Grace", 45)]

print(sorted(people, key=lambda person: person[1], reverse=True))

[('Grace', 45), ('Alan', 41), ('Ada', 36)]


A lambda is limited to a single expression, so anything with an if/else chain or a loop needs a proper `def`. Use a lambda when the function is genuinely a one-liner and is only needed once.

Lambdas come back as soon as you meet `.apply()` in Part 3, because the data libraries let you hand one function to a whole column of data at a time.

> **&#9998; Exercise.** Using `sorted()` and a lambda, sort the list `["Sheffield", "Leeds", "York", "Manchester"]` by the length of each name, shortest first. Then sort it by the last letter of each name.

In [ ]:
# your answer here
cities = ["Sheffield", "Leeds", "York", "Manchester"]

# Sort by length, shortest first
by_length = sorted(cities, key=lambda name: len(name))
print(by_length)
# ['York', 'Leeds', 'Sheffield', 'Manchester']

# Sort by last letter
by_last_letter = sorted(cities, key=lambda name: name[-1])
print(by_last_letter)
# ['Sheffield', 'York', 'Manchester', 'Leeds']

['York', 'Leeds', 'Sheffield', 'Manchester']
['Sheffield', 'York', 'Manchester', 'Leeds']


## 6. Doing the same thing to every item: map, filter and reduce

These three functions express three common jobs compactly: transform every item, keep some items, and boil a collection down to a single value. You will meet them in other people's code, and they pair naturally with lambdas.

### 6.1 map: transform every item

map() applies a function to every item of a collection. It hands back a lazy iterator rather than a list, so we usually wrap it in `list()` to see the result:

In [ ]:
numbers = [1, 2, 3, 4, 5]
squares = list(map(lambda x: x ** 2, numbers))
print(squares)

[1, 4, 9, 16, 25]


The function does not have to be a lambda. Any function will do:

In [ ]:
words = ["banana", "fig", "cherry"]
print(list(map(len, words)))

[6, 3, 6]


### 6.2 filter: keep some items

filter() keeps only the items for which the function returns True:

In [ ]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]
evens = list(filter(lambda x: x % 2 == 0, numbers))
print(evens)

[2, 4, 6, 8]


### 6.3 reduce: boil down to one value

reduce() applies a two-argument function cumulatively: to the first two items, then to that result and the third item, and so on, until one value is left. It lives in the `functools` module, so it has to be imported:

In [ ]:
from functools import reduce

print(reduce(lambda x, y: x + y, [1, 2, 3, 4, 5]))

15


Which, for adding up, is a long way of writing something Python already gives you:

In [ ]:
print(sum([1, 2, 3, 4, 5]))

15


That is worth saying plainly. These three are elegant, but in modern Python there is usually a more readable alternative, and for data work the tools in the next section do the same jobs faster and more clearly. A list comprehension covers most of what map and filter do:

In [ ]:
print([x ** 2 for x in [1, 2, 3, 4, 5]])          # instead of map
print([x for x in range(1, 10) if x % 2 == 0])    # instead of filter

[1, 4, 9, 16, 25]
[2, 4, 6, 8]


So: know them, recognise them when you read them, and reach for the clearer option when you write your own.

> **&#9998; Exercise.** Given a sentence, use `map()` to produce a list of the length of every word in it. Then do the same with a list comprehension and satisfy yourself that they agree.

In [ ]:
# your answer here
sentence = "The is a testing sentence"
words = sentence.split()

# Using map()
lengths_map = list(map(len, words))

# Using a list comprehension
lengths_comp = [len(word) for word in words]

print(lengths_map)                  # [3, 5, 5, 3, 5, 4, 3, 4, 3]
print(lengths_comp)                 # [3, 5, 5, 3, 5, 4, 3, 4, 3]
print(lengths_map == lengths_comp)  # True

[3, 2, 1, 7, 8]
[3, 2, 1, 7, 8]
True


> **&#9998; Exercise.** Given a list of words, use `filter()` to keep only those longer than four characters.

In [ ]:
# your answer here
words = ["The","fox", "jumps", "over", "the", "lazy", "dog"]

long_words = list(filter(lambda word: len(word) > 4, words))
print(long_words)  # ['quick', 'brown', 'jumps']

['jumps']


## 7. From lists to NumPy arrays



### 7.1 Why a list is not enough

Suppose you have four measurements and want to double each one. With a list, the obvious thing does not do what you would hope:

In [ ]:
print([1.2, 2.3, 0.2, 1.1] * 2)

[1.2, 2.3, 0.2, 1.1, 1.2, 2.3, 0.2, 1.1]


Multiplying a list repeats the whole list, which is almost never what you want when the numbers are data. You could write a loop, and for four values that is fine. But real datasets have millions of values, and a Python loop over a million numbers is slow to write and slow to run.

### 7.2 Arrays

NumPy is the package that gives Python fast numerical arrays. An array holds values that are all of the same type, and, this is the important part, an operation applies to **every element at once**. That idea is called vectorisation and it underpins everything we do from here on.

NumPy is already installed in Colab, so we only need to load it. The convention, which you will see in every piece of Python data code you ever read, is to import it under the short name `np`:

In [ ]:
import numpy as np

From now on `np` is our nickname for the package, and we reach its functions with a dot. Put the import in a cell near the top of any notebook that uses it, so that Run all works.

We make an array out of a list:

In [ ]:
scores = np.array([1.0, 2.3, 0.2, 1.1])
scores

array([1. , 2.3, 0.2, 1.1])

And now the arithmetic does what you wanted in the first place:

In [ ]:
print(scores * 2)
print(scores + 10)
print(scores ** 2)

[2.  4.6 0.4 2.2]
[11.  12.3 10.2 11.1]
[1.   5.29 0.04 1.21]


No loop, one line, and for large datasets many times faster. Two useful things an array knows about itself are `dtype`, the type of the values it holds, and `shape`, its dimensions. There are no brackets after these, because they are properties, not functions:

In [ ]:
print(scores.dtype)
print(scores.shape)
print(len(scores))

float64
(4,)
4


Indexing and slicing work exactly as they did for lists, still counting from 0.

### 7.3 Asking questions of a whole array

Comparisons are vectorised too. Compare an array with a value and you get back an array of True/False answers, one for each element:

In [ ]:
print(scores > 2)

[False  True False False]


To combine two conditions across an array, use the symbols `&` and `|` rather than the words `and` and `or`, and give each condition its own brackets. (The words ask a single true-or-false question; the symbols ask the question of every element.)

In [ ]:
print((scores > 1) & (scores <= 4))

[False  True False  True]


Now the payoff. Put a condition inside the square brackets and you get back just the elements where it was true. This is called boolean indexing, and it is how nearly all data filtering is done in Python:

In [ ]:
print(scores[scores > 2])

[2.3]


### 7.4 Summarising an array

NumPy provides the summary functions. Many are available both as a method on the array and as a function in the package; a few, including median and percentile, come only as a function:

In [ ]:
print(scores.mean(), scores.sum(), scores.min(), scores.max())

1.15 4.6 0.2 2.3


In [ ]:
print(np.median(scores))
print(np.percentile(scores, 25), np.percentile(scores, 75))

1.05
0.8 1.4


> **Note:** keep the `print()`. Without it, a single number taken out of an array displays as something like `np.float64(1.15)`. That is just NumPy pointing out that the value is one of its own numbers rather than a plain Python float. It behaves exactly like a float.

Sorting, and adding values on the end:

In [ ]:
print(np.sort(scores))
print(np.sort(scores)[::-1])     # sort, then reverse

[0.2 1.  1.1 2.3]
[2.3 1.1 1.  0.2]


In [ ]:
scores = np.append(scores, [5.2, 7.3])
print(scores)
print(scores.mean(), len(scores))

[1.  2.3 0.2 1.1 5.2 7.3]
2.85 6


Note that `np.append()` hands back a **new** array rather than changing the original, which is why we assign the result back. That is the opposite of a list's `.append()` method.

An array can hold text, but all of its elements must be of the same type:

In [ ]:
print(np.array(["Tom", "Deep", "Harry"]).dtype)

<U5


(The odd-looking <U5 simply means “text of up to 5 characters”.)

### 7.5 Two-dimensional arrays

An array can have more than one dimension. A two-dimensional array is a table of rows and columns, the natural home for something like a medal table. Give np.array() a list of lists, one per row:

In [ ]:
table = np.array([[46, 37, 38],      # USA
                  [27, 23, 17],      # GBR
                  [26, 18, 26]])     # CHN
table

array([[46, 37, 38],
       [27, 23, 17],
       [26, 18, 26]])

Its `shape` now reports rows and columns:

In [ ]:
print(table.shape)

(3, 3)


To reach an element you give a row and a column, separated by a comma. A bare colon means “all of them along this dimension”:

In [ ]:
print(table[0, 1])     # first row, second column
print(table[0, :])     # the whole first row
print(table[:, 0])     # the whole first column

37
[46 37 38]
[46 27 26]


The really useful part is summarising along one dimension, using the `axis` argument. Read `axis=1` as “collapse the columns, giving me one number per row”, and `axis=0` as “collapse the rows, giving me one number per column”:

In [ ]:
print(table.sum(axis=1))     # total medals for each country
print(table.sum(axis=0))     # total of each medal type
print(table.sum())           # everything

[121  67  70]
[99 78 81]
258


The same pattern works for `mean()`, `min()`, `max()` and `std()`.

One obvious thing is missing from that table: the country names. An array is all one type and is addressed by number only, so there is nowhere to put them. That is exactly the gap the pandas DataFrame fills, and it is what Part 3 is built on.

> **&#9998; Exercise.** Build the array above, then use `sum(axis=1)` to get each country's total and print the largest of those totals.

In [ ]:
# your answer here
totals = table.sum(axis=1)
print(totals)        # [121  67  70]
print(totals.max())  # 121

[121  67  70]
121


> **&#9998; Exercise.** Create an array of the numbers 1 to 10 with `np.array(range(1, 11))`. Then, without writing a loop: print the array doubled, print only the values greater than 5, and print the mean of those values.

In [ ]:
# your answer here
numbers = np.array(range(1, 11))

print(numbers * 2)

big = numbers[numbers > 5]
print(big)

print(big.mean())

[ 2  4  6  8 10 12 14 16 18 20]
[ 6  7  8  9 10]
8.0


> **&#9998; Exercise.** Convert this list of temperatures in Celsius to Fahrenheit in one line, using the formula F = C × 9/5 + 32: `[12.1, 15.6, 9.8, 21.3, 18.0]`. Then print how many of them are above 60°F.

In [ ]:
# your answer here

celsius = np.array([12.1, 15.6, 9.8, 21.3, 18.0])

fahrenheit = celsius * 9/5 + 32
print(fahrenheit)

print((fahrenheit > 60).sum())

[53.78 60.08 49.64 70.34 64.4 ]
3


That is the end of Part 2. You can now make decisions, repeat work, package code into functions, and operate on whole sets of numbers at once. In Part 3 we put those skills to work on real data: tables with named columns, files, charts, and a live dataset from the World Bank.

Useful resources

* The official Python tutorial, sections 4 and 5 (control flow and functions): [https://docs.python.org/3/tutorial/controlflow.html](https://docs.python.org/3/tutorial/controlflow.html)
* NumPy for absolute beginners: [https://numpy.org/doc/stable/user/absolute_beginners.html](https://numpy.org/doc/stable/user/absolute_beginners.html)
* Python for Data Analysis by Wes McKinney, Chapters 3 and 4: [https://wesmckinney.com/book/](https://wesmckinney.com/book/)
* Python Data Science Handbook by Jake VanderPlas, Chapter 2: [https://jakevdp.github.io/PythonDataScienceHandbook/](https://jakevdp.github.io/PythonDataScienceHandbook/)